proxy/coco/src/config.rs

```diff
@@ -133,22 +135,39 @@ impl discovery::Discovery for StreamDiscovery {
     type Stream = mpsc::Receiver<(PeerId, Vec<SocketAddr>)>;
 
     fn discover(mut self) -> Self::Stream {
+        let last_seeds: Arc<Mutex<Vec<seed::Seed>>> = Arc::new(Mutex::new(vec![]));
         let (mut sender, receiver) = mpsc::channel(1024);
 
+        {
+            let last_seeds = last_seeds.clone();
+            let mut sender = sender.clone();
+            tokio::spawn(async move {
```

Comment:
Is there a reason why we poll the streams explicitly instead of using stream combinators like `flat_map` and `scan` to return the result stream? Using the combinators would be more concise and we wouldn’t need to worry about the task if the consumer terminates.

Reply:
We have to spawn this task explicitely as we need to return from the method. Not quite what you explicitely would change heere.

Reply:
I’d suggest the following implementation
~~~rust
fn discover(mut self) -> Self::Stream {
    let interval = tokio::time::interval(Duration::from_secs(5)).map(Either::Left);
    let seed_updates = self.seeds_receiver.map(Either::Right);

    futures::stream::select(interval, seed_updates)
        .scan(vec![], |last_seeds, item| {
            let seeds = match item {
                Either::Left(_) => last_seeds.clone(),
                Either::Right(seeds) => {
                    if *last_seeds == seeds {
                        vec![]
                    } else {
                        *last_seeds = seeds;
                        last_seeds.clone()
                    }
                },
            };
            futures::future::ready(Some(seeds))
        })
        .flat_map(|seeds| {
            futures::stream::iter(
                seeds
                    .into_iter()
                    .map(|seed| (seed.peer_id, vec![seed.addr])),
            )
        })
        .boxed()
}
~~~

Reply:
Elegant solution!